### PR TITLE
coprocessor: decode_json using BufferReader

### PR DESCRIPTION
**Pull request by niedhui**
Signed-off-by: niedhui <niedhui@gmail.com>

###  What have you changed?

this PR use BufferReader of codec component to decode decimal
and remove the usage of `tikv_util::codec::read_slice`

###  What is the type of the changes?

Improvement 

###  How is the PR tested?

unit test

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?

No

###  Does this PR affect `tidb-ansible`?

No

###  Benchmark result if necessary (optional)

I ran benchmark on my MBP using
`cargo bench --color=always --package tikv --bench misc coprocessor::codec::mysql::json::bench_decode_binary -- --exact`
on master: 
```
test coprocessor::codec::mysql::json::bench_decode_binary   ... bench:   9,273,552 ns/iter (+/- 1,273,623)
```
on  this PR
```
test coprocessor::codec::mysql::json::bench_decode_binary   ... bench:   7,743,437 ns/iter (+/- 744,749)
```
If I'm using the wrong benchmark, please let me know

### CODE REVIEW

**Comment by siddontang** [comment]
thanks @niedhui 

seem the performance improvement is introduced by removing `read_slice`, right?

**Comment by niedhui** [comment]
@siddontang  there's little performance improvements by removing `read_slice`, but not that much. 

@breeswish 
I'm a little confusing when trying to find out which is the critical path for the improvements. If you had time, please help to review the code first, in case I got the wrong bench result, thanks.

when I take the `impl Json` back, even only delegate the `decode` to `JsonDecoder`,  the performance is down back...

```
impl Json {
    // `decode` decodes value encoded by `encode_json` before.
    pub fn decode(buf: &mut &[u8]) -> Result<Json> {
        buf.decode_json()
    }
}
```
the result:
```
test coprocessor::codec::mysql::json::bench_decode_binary_buffer_reader  ... bench:   7,788,137 ns/iter (+/- 608,922)
test coprocessor::codec::mysql::json::bench_decode_binary_impl_json      ... bench:   9,223,332 ns/iter (+/- 784,316)
```

File: components/tidb_query/src/codec/mysql/json/binary.rs [hunk 1]
```diff
@@ -219,125 +219,141 @@ pub trait JsonEncoder: NumberEncoder {
...
 
-    fn decode_item(
-        buf: &mut BytesSlice<'_>,
-        values_data: &[u8],
-        data_start_position: u32,
-    ) -> Result<Json> {
-        let entry = read_slice(buf, VALUE_ENTRY_LEN)?;
-        let code = entry[0];
+    fn decode_json_item(&mut self, values_data: &[u8], data_start_position: u32) -> Result<Json> {
+        let code = self.read_u8()?;
```
[the remark this answers is not in the document]
Comment (breeswish) [reply components/tidb_query/src/codec/mysql/json/binary.rs]: How about:

```rust
let entry = self.read_bytes(VALUE_ENTRY_LEN)?;
let code = entry.read_u8()?;
...
TYPE_CODE_LITERAL => entry.decode_json_literal(),
...
```

In this way, code difference can be minimized.

[the remark this answers is not in the document]
Comment (niedhui) [reply components/tidb_query/src/codec/mysql/json/binary.rs]: done

File: components/tidb_query/src/codec/mysql/json/binary.rs [hunk 2]
```diff
@@ -219,125 +219,133 @@ pub trait JsonEncoder: NumberEncoder {
...
             return Ok(Json::Object(obj));
         }
+        let buf = self.bytes();
+        if buf.len() < left_size {
+            return Err(Error::unexpected_eof());
+        }
         // key_entries
         let key_entries_len = KEY_ENTRY_LEN * element_count;
-        let mut key_entries_data = read_slice(buf, key_entries_len)?; //&data[0..key_entries_len];
+        let mut key_entries_data = &buf[..key_entries_len]; //&data[0..key_entries_len];
```
[the remark this answers is not in the document]
Comment (niedhui) [reply components/tidb_query/src/codec/mysql/json/binary.rs]: done

File: components/tidb_query/src/codec/mysql/json/binary.rs [hunk 3]
```diff
@@ -219,125 +219,133 @@ pub trait JsonEncoder: NumberEncoder {
...
-        let left_size = total_size as usize - ELEMENT_COUNT_LEN - SIZE_LEN;
+        let left_size = total_size - ELEMENT_COUNT_LEN - SIZE_LEN;
+        let buf = self.bytes();
+        if buf.len() < left_size {
+            return Err(Error::unexpected_eof());
+        }
         let value_entries_len = VALUE_ENTRY_LEN * element_count;
-        let mut value_entries_data = read_slice(buf, value_entries_len)?;
-        let values_data = read_slice(buf, left_size - value_entries_len)?;
+        let mut value_entries_data = &buf[..value_entries_len];
```
[the remark this answers is not in the document]
Comment (niedhui) [reply components/tidb_query/src/codec/mysql/json/binary.rs]: done

File: components/tidb_query/src/codec/mysql/json/binary.rs [hunk 4]
```diff
@@ -219,125 +219,128 @@ pub trait JsonEncoder: NumberEncoder {
...
-        let mut data = read_slice(buf, left_size - key_entries_len - value_entries_len)?;
+        let (mut value_entries_data, mut data) = buf.split_at(value_entries_len);
         for _ in 0..element_count {
-            let key_real_offset = number::decode_u32_le(&mut key_entries_data)?;
-            let key_len = number::decode_u16_le(&mut key_entries_data)?;
-            let key_data = read_slice(&mut data, key_len as usize)?;
+            let key_real_offset = key_entries_data.read_u32_le()?;
+            let key_len = key_entries_data.read_u16_le()? as usize;
+            let key_data = data.read_bytes(key_len)?;
             let key = String::from(str::from_utf8(key_data).unwrap());
```
[the remark this answers is not in the document]
Comment (niedhui) [reply components/tidb_query/src/codec/mysql/json/binary.rs]: `unwrap` removed

**Pull request merged**